src/app/components/RuntimeEvents/RuntimeEventDetails.tsx

```diff
@@ -256,15 +268,23 @@ export const RuntimeEventDetails: FC<{
             <MaybeEventErrorLine event={event} />
             <dt>{t('common.from')}</dt>
             <dd>
-              <AccountLink address={event.body.from} scope={scope} />
-              {addressSwitchOption === AddressSwitchOption.Oasis && (
-                <CopyToClipboard value={event.body.from} />
-              )}
+              <AccountLinkWithAddressSwitch
+                address={event.body.from_eth ?? event.body.from}
+                scope={scope}
+                addressSwitchOption={addressSwitchOption}
+                ethAddress={event.body.from_eth}
+                oasisAddress={event.body.from}
+              />
             </dd>
             <dt>{t('common.to')}</dt>
             <dd>
-              <AccountLink address={event.body.to} scope={scope} />
-              {addressSwitchOption === AddressSwitchOption.Oasis && <CopyToClipboard value={event.body.to} />}
+              <AccountLinkWithAddressSwitch
+                address={event.body.to_eth ?? event.body.to}
```

Comment:
I would remove `address` from props

Reply:
Yeah, I have been messing with TypeScript, didn't figure it out quite yet, how to do that. Will try again next week 😅 But did try to implement it that way.

Reply:
https://github.com/oasisprotocol/explorer/blob/cabd21b96ec7787c4ce6511f8cfd8850cc3e1f77/src/app/components/Account/AccountLinkWithAddressSwitch.tsx#L1-L5

This is the best I came up with, if you have any ideas, how to do it better, would appreciate, if you would take a look

Reply:
I think this is better:
```tsx
export const withAddressSwitch = <T extends WrappedComponentBaseProps>(
  Component: FunctionComponent<T>,
): FunctionComponent<WithAddressSwitchProps & Omit<T, 'address'>> => {
  return props => {
    const { addressSwitchOption, ethAddress, oasisAddress, ...restProps } = props

    const addressesByType = {
      [AddressSwitchOption.Oasis]: oasisAddress,
      [AddressSwitchOption.ETH]: ethAddress,
    }
    const hasAddressOfExpectedType = !!addressesByType[addressSwitchOption]
    const displayedAddress = addressesByType[addressSwitchOption] ?? ethAddress ?? oasisAddress

    return (
      <>
        <Tooltip
          arrow
          placement="top"
          title={
            <Box sx={{ display: 'flex', alignItems: 'center', gap: 2 }}>
              <InfoIcon />
              {/* TODO: Replace with translations, text to be decided?, should be probably an input to this generic withAddressSwitch */}
              {hasAddressOfExpectedType
                ? oasisAddress
                  ? 'Oasis address'
                  : 'Ethereum address'
                : 'Address not available in the expected type'}
            </Box>
          }
        >
          <Box>
            {/* Ignore type error: I think its because Component could require props like "oasisAddress" and restProps would not provide them */}
            <Component {...(restProps as any as T)} address={displayedAddress} />
          </Box>
        </Tooltip>
        {displayedAddress && <CopyToClipboard value={displayedAddress} />}
      </>
    )
  }
}
```
```tsx
import { AccountLink, AccountLinkProps } from './AccountLink'
import { withAddressSwitch } from './withAddressSwitch'

export const AccountLinkWithAddressSwitch = withAddressSwitch(AccountLink)
```